frontend/src/views/Room.jsx

```diff
@@ -1,19 +1,168 @@
+  // this array will hold data such as name foor the purpose of
+  // having a participants name list, etc.
+
+  // It's possible that visibleParitcipants will eventually not need to be in the state,
+  // if it ends up being a component property passed in by the parent component.
+  // for now it needs to be in the state so that the "Add Participant" test button can work
+  const rowsLimit = { xs: 2, sm: 3, md: 4 };
+  const tilesPerRowLimit = { xs: 2, sm: 3, md: 6 };
+  // it's assumed that we'll get a maximum of rowsLimit*tilesPerRowLimit visibleParticipants
+  // if this precondition isn't true then the video grid can't be expected to render properly
+  const calculateTilesPerRow = (screenSize) => {
+    // screenSize should be either 'xs', 'sm' or 'md'
+    /* This function attempts to distribute the tiles between
+    rows as evenly as possible. It does so while trying to mantain
+    the amount of tiles per row bigger or equal that the total amount
+    of rows, as this looks better than the opposite relation. */
+    const tilesAmount = remoteStreams.length;
+    let rows = 1;
+    while (rows <= rowsLimit[screenSize]) {
+      const tilesPerRow = Math.ceil(tilesAmount / (rows + 1));
+      if (
+        tilesPerRow >= rows + 1
+      ) {
+        rows += 1;
+      } else {
+        break;
+      }
+    }
+    /*
+    TODO? fix the "7" case (ask Nico Reyes)
+    const finalTilesPerRow = Math.ceil(tilesAmount / (rows));
+    const tilesInLastRow = finalTilesPerRow % rows;
+    if (finalTilesPerRow - (tilesInLastRow) >= 2) {
+      rows -= 1;
+    }
+    */
+    return Math.min(
+      tilesPerRowLimit[screenSize],
+      Math.ceil(tilesAmount / rows),
+    );
+  };
+  const tilesPerRow = {
+    xs: calculateTilesPerRow('xs'),
+    sm: calculateTilesPerRow('sm'),
+    md: calculateTilesPerRow('md'),
+  };
+  // initialize room
+  useEffect(() => {
+    const leaveRoom = async () => {
+      if (room) {
+        await userParticipant.unpublishAllTracks(); // also stops them
+        await room.leave();
+      }
+    };
+    const subscribeToRemoteStreams = async (r) => {
+      // subscribe ta all remote participants for testing purposes
+      const { remoteParticipants } = r;
+      const rps = Array.from(remoteParticipants.values());
+      await Promise.all(rps.map(async (rp) => {
+        await rp.subscribe();
+      }));
+      // console.log('subscribed to remote participant(s)');
+    };
+    const joinRoom = async () => {
+      const JWT = await roomJWTprovider(roomId, null, null, () => { setRoomNotFound(true); });
+      const newRoom = new WebRoom(JWT);
+      const newParticipant = await newRoom.join();
+
+      newRoom.on('ParticipantTrackSubscribed', (remoteParticipant, track) => {
+        const stream = new MediaStream();
+        stream.addTrack(track.mediaStreamTrack);
+        const streamObj = { stream, participantId: remoteParticipant.id };
+        setRemoteStreams([...remoteStreams, streamObj]);
+      });
```

Comment:
I'm not sure this is working... Also, here we create a stream for each track, rather than adding both audio and video tracks to the same stream, like we would with:

```js
tracks.forEach((track) => stream.addTrack(track.mediaStreamTrack));
```